{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get's intracohort structural variant count based on \n",
    "# 50% reciprocal overlap between structural variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "svs = pd.read_csv('output/merged.PennCNV.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop over every chrom and svtype\n",
    "chromosomes = list(svs.Chromosome.unique())\n",
    "svtypes = ['dup', 'del']\n",
    "\n",
    "for chrom in chromosomes:\n",
    "    for svtype in svtypes:\n",
    "        # get all structural variants in chromosome of the svtype\n",
    "        sub_svs = svs[(svs.Chromosome == chrom) & (svs.Type == svtype)].copy()\n",
    "\n",
    "        # create numpy structures for end, start, and svlength\n",
    "        # (numpy is faster than pandas)\n",
    "        chrom_ends = sub_svs['End'].to_numpy()\n",
    "        chrom_starts = sub_svs['Start'].to_numpy()\n",
    "        chrom_svlengths = sub_svs['Length'].to_numpy()\n",
    "\n",
    "        # iterate over all structural variants in chrom\n",
    "        # and compare against other structural variants in chrom\n",
    "        for i, row in sub_svs.iterrows():\n",
    "            start = row['Start']\n",
    "            end = row['End']\n",
    "            length = row['Length']\n",
    "\n",
    "            min_end   = np.minimum(chrom_ends, end)\n",
    "            max_start = np.maximum(chrom_starts, start)\n",
    "            max_length = np.maximum(chrom_svlengths, length)\n",
    "\n",
    "            overlap = (min_end - max_start) > .5 * max_length\n",
    "\n",
    "            intra_cohort_count = overlap.sum()\n",
    "            svs.at[i, 'intracohort_count'] = intra_cohort_count\n",
    "            \n",
    "svs['intracohort_count'] = svs['intracohort_count'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "svs.to_csv('output/merged.PennCNV.intracohort.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>NumSNPs</th>\n",
       "      <th>Length</th>\n",
       "      <th>Type</th>\n",
       "      <th>StartSNP</th>\n",
       "      <th>EndSNP</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>%SD</th>\n",
       "      <th>%CenTel</th>\n",
       "      <th>Genes</th>\n",
       "      <th>GeneList</th>\n",
       "      <th>intracohort_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>752721</td>\n",
       "      <td>854250</td>\n",
       "      <td>8</td>\n",
       "      <td>101530</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>rs7537756</td>\n",
       "      <td>SG172</td>\n",
       "      <td>0.9770</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>5</td>\n",
       "      <td>FAM87B,LINC00115,LINC01128,FAM41C,LOC100130417</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>959842</td>\n",
       "      <td>1026707</td>\n",
       "      <td>16</td>\n",
       "      <td>66866</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs2710888</td>\n",
       "      <td>rs4074137</td>\n",
       "      <td>SG467</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.6006</td>\n",
       "      <td>3</td>\n",
       "      <td>AGRN,RNF223,C1orf159</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>959842</td>\n",
       "      <td>1040026</td>\n",
       "      <td>19</td>\n",
       "      <td>80185</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs2710888</td>\n",
       "      <td>rs6671356</td>\n",
       "      <td>SG469</td>\n",
       "      <td>0.0089</td>\n",
       "      <td>0.5008</td>\n",
       "      <td>3</td>\n",
       "      <td>AGRN,RNF223,C1orf159</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>962210</td>\n",
       "      <td>1025301</td>\n",
       "      <td>13</td>\n",
       "      <td>63092</td>\n",
       "      <td>del</td>\n",
       "      <td>rs3128126</td>\n",
       "      <td>rs9442400</td>\n",
       "      <td>SG045</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5990</td>\n",
       "      <td>3</td>\n",
       "      <td>AGRN,RNF223,C1orf159</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>962210</td>\n",
       "      <td>1026707</td>\n",
       "      <td>15</td>\n",
       "      <td>64497</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs3128126</td>\n",
       "      <td>rs4074137</td>\n",
       "      <td>SG404</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5859</td>\n",
       "      <td>3</td>\n",
       "      <td>AGRN,RNF223,C1orf159</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2603</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149444023</td>\n",
       "      <td>149593953</td>\n",
       "      <td>23</td>\n",
       "      <td>149931</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs9723612</td>\n",
       "      <td>rs683913</td>\n",
       "      <td>SG334</td>\n",
       "      <td>0.3775</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>MAMLD1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2604</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151170560</td>\n",
       "      <td>151239285</td>\n",
       "      <td>6</td>\n",
       "      <td>68726</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs11798839</td>\n",
       "      <td>rs1388517</td>\n",
       "      <td>SG399</td>\n",
       "      <td>0.0399</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2605</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152228236</td>\n",
       "      <td>152596189</td>\n",
       "      <td>20</td>\n",
       "      <td>367954</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs1045059</td>\n",
       "      <td>rs12841019</td>\n",
       "      <td>SG482</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3</td>\n",
       "      <td>PNMA3,PNMA6A,MAGEA1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2606</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152480761</td>\n",
       "      <td>152626413</td>\n",
       "      <td>23</td>\n",
       "      <td>145653</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs6627761</td>\n",
       "      <td>rs12841273</td>\n",
       "      <td>SG177</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2</td>\n",
       "      <td>MAGEA1,ZNF275</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2607</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153168154</td>\n",
       "      <td>153231748</td>\n",
       "      <td>15</td>\n",
       "      <td>63595</td>\n",
       "      <td>dup</td>\n",
       "      <td>rs12559136</td>\n",
       "      <td>rs12011769</td>\n",
       "      <td>SG270</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>5</td>\n",
       "      <td>AVPR2,ARHGAP4,NAA10,RENBP,HCFC1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2608 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome      Start        End  NumSNPs  Length Type    StartSNP  \\\n",
       "0          chr1     752721     854250        8  101530  dup   rs3131972   \n",
       "1          chr1     959842    1026707       16   66866  dup   rs2710888   \n",
       "2          chr1     959842    1040026       19   80185  dup   rs2710888   \n",
       "3          chr1     962210    1025301       13   63092  del   rs3128126   \n",
       "4          chr1     962210    1026707       15   64497  dup   rs3128126   \n",
       "...         ...        ...        ...      ...     ...  ...         ...   \n",
       "2603       chrX  149444023  149593953       23  149931  dup   rs9723612   \n",
       "2604       chrX  151170560  151239285        6   68726  dup  rs11798839   \n",
       "2605       chrX  152228236  152596189       20  367954  dup   rs1045059   \n",
       "2606       chrX  152480761  152626413       23  145653  dup   rs6627761   \n",
       "2607       chrX  153168154  153231748       15   63595  dup  rs12559136   \n",
       "\n",
       "          EndSNP PatientID     %SD  %CenTel  Genes  \\\n",
       "0      rs7537756     SG172  0.9770   1.0000      5   \n",
       "1      rs4074137     SG467  0.0107   0.6006      3   \n",
       "2      rs6671356     SG469  0.0089   0.5008      3   \n",
       "3      rs9442400     SG045  0.0000   0.5990      3   \n",
       "4      rs4074137     SG404  0.0000   0.5859      3   \n",
       "...          ...       ...     ...      ...    ...   \n",
       "2603    rs683913     SG334  0.3775   0.0000      1   \n",
       "2604   rs1388517     SG399  0.0399   0.0000      0   \n",
       "2605  rs12841019     SG482  0.0049   0.0000      3   \n",
       "2606  rs12841273     SG177  0.0036   0.0000      2   \n",
       "2607  rs12011769     SG270  0.0000   0.0000      5   \n",
       "\n",
       "                                            GeneList  intracohort_count  \n",
       "0     FAM87B,LINC00115,LINC01128,FAM41C,LOC100130417                  1  \n",
       "1                               AGRN,RNF223,C1orf159                  5  \n",
       "2                               AGRN,RNF223,C1orf159                 14  \n",
       "3                               AGRN,RNF223,C1orf159                  1  \n",
       "4                               AGRN,RNF223,C1orf159                  5  \n",
       "...                                              ...                ...  \n",
       "2603                                          MAMLD1                  1  \n",
       "2604                                             NaN                  1  \n",
       "2605                             PNMA3,PNMA6A,MAGEA1                  1  \n",
       "2606                                   MAGEA1,ZNF275                  1  \n",
       "2607                 AVPR2,ARHGAP4,NAA10,RENBP,HCFC1                  1  \n",
       "\n",
       "[2608 rows x 14 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
