# Description
This directory holds the preprocessing scripts for raw data analyzed in the manuscript. It is further sub-divided into folders.

# Phenotype
*0_Child_Domains_Jul16.ipynb*: Preparation of different phenotypic domains based on observed phenotypes

# Genotype
## Genome alignment and SNV calls
*0_gatk_pipe_array.sh*: Trimming, aligning, filtering, duplicate removal and SNV calling using gatk's best practices pipeline.

## SNV annotation
*1_list_of_files.py*: 

*2_filter_quality.sh*:

*3_annovar_genes.sh*:

*4_filter_exonic.sh*:

*5_annotate_gnomad.sh*:

*6_filter_pop_freq.sh*:

*7_annotate_cadd.sh*:

*8_vcf2table.sh*:

*9_combine_tables.sh*:

*10_filter_variant_types.py*:

*11_intracohort_filter.py*:

*12_gencode_gene_ids.py*:

*13_louef_scores.py*:

*14_omim_annotations.py*:

*15_finalize_variants.sh*:
